[VIP-2463] Add debug data export button to settings panel

shekharnwagh opened the pull request:
> Linear - [VIP-2463](https://linear.app/a8c/issue/VIP-2463)

Adds a new button in the RTC settings panel that exports the following debug data JSON with following structure -

```typescript
interface YDocDebugData {
	doc: {
		state: {
			version: number;
		};
		// YJS representation of Gutenberg entity
		document: {
			[key: string]: unknown;
		};
	};
	clients: {
		[clientId: string]: {
			id: {
				client: number;
				clock: number;
			};
			length: number;
			origin: string | null;
			rightOrigin: string | null;
			// YJS parent object
			parent: Record<string, unknown>;
			// key in the parent object for the current item
			parentSub: string;
			redone: Record<string, unknown> | null;
			content: Record<string, unknown>;
			info: number;
			// YJS left item reference
			left: Record<string, unknown> | null;
			// YJS right item reference
			right: Record<string, unknown> | null;
		}[];
	};
	// Map of the YJS client ID to the WordPress user as YJS client ID is re-generated on each refresh
	userMap: {
		[clientId: string]: {
			wpUserId: number;
			name: string;
			email: string;
		};
	};
}
```

This would be useful in understanding how the Post blocks/attributes are updated in a collaborative session by participants.

### Screenshots

<img width="1449" height="859" alt="Screenshot 2025-11-25 at 10 38 29 PM" src="https://github.com/user-attachments/assets/4ea5be33-53a3-4286-9040-bed1fd3148fc" />

shekharnwagh committed:
Add debug data export button to settings panel

Adds a new button in the RTC settings panel that exports YDoc client
data (YDoc.store.clients) to a JSON file for debugging collaboration
issues.

shekharnwagh committed:
Include YDoc content in debug data export

  Changed getYDocClients to getDebugData that returns both document
  content and client structure information. The export now includes:
  - doc: YDoc shared types serialized to JSON
  - clients: Internal client structure from doc.store.clients

shekharnwagh committed:
strip nested field to cut down debug data size

chriszarate commented on src/components/debug-data-export-button.tsx:
```diff
@@ -0,0 +1,38 @@
+import { AwarenessManager } from '@/awareness-manager';
+
+export function DebugDataExportButton() {
+	const handleDownload = () => {
```
this function can be in outer scope

chriszarate commented on src/awareness-manager.ts:
```diff
@@ -93,6 +125,50 @@ export class AwarenessManager {
 	}
 
+	public static getDebugData(): YDocDebugData | null {
+		const logger = new Logger( 'awareness-manager' );
```
Can we create this once in outer scope and remove `this.logger`?

chriszarate commented on src/awareness-manager.ts:
```diff
@@ -93,6 +125,50 @@ export class AwarenessManager {
+		const docData: Record< string, unknown > = {};
+		ydoc.share.forEach( ( value, key ) => {
+			docData[ key ] = value.toJSON();
+		} );
```
```suggestion
		const docData: Record< string, unknown > = Object.fromEntries(
			Object.entries( ydoc.share ).forEach( ( [ value, key ] ) => [ key, value.toJSON() ] )
		);
```

chriszarate committed:
Keep all seen users in userMap

chriszarate commented on src/awareness-manager.ts:
```diff
@@ -111,6 +187,39 @@ export class AwarenessManager {
+	 * Persist a user client mapping in the YDoc.
+	 * Stores Yjs client ID -> WordPress user info mapping for debugging.
+	 */
+	private persistUserClientsMap( yjsClientId: number, userInfo: UserInfo ): void {
```
Instead of storing these on the Yjs state map, I think each client can just keep an in memory map of all the clients they have seen. Not perfect (misses users that the current client has not seen), but perhaps good enough for now

https://github.com/Automattic/vip-real-time-collaboration/commit/e3c6eb2698a6c4af5f905a40845bd6a98902663c

shekharnwagh committed:
Use awareness store for debug data user mapping

Moves Yjs client-to-user mapping from YDoc state persistence to
awareness store's in-memory tracking. We had issues with getting that
to sync correctly. This is simpler to start with than looking into why
it's not working. This has the trade-off of only including users seen
after current client joined, but this is acceptable as first iteration.

shekharnwagh committed:
move handleDebugDataDownload to global scope to avoid re-defining it in the component

shekharnwagh committed:
use existing static logger instance

shekharnwagh committed:
refactor serializing ydoc for debug data

Co-authored-by: Chris Zarate <chris.zarate@automattic.com>

chriszarate committed:
Settings UI cleanup

chriszarate committed:
Return empty debug data when there is no awareness instance

chriszarate reviewed (approved):
I pushed a couple of cleanup commits. Looks great!

shekharnwagh committed:
Merge branch 'trunk' into try/diagnostics-yjs

ingeniumed commented on src/components/rtc-settings-panel.tsx:
```diff
@@ -143,10 +144,6 @@ export function RTCSettingsPanel() {
 						{ __( 'Notifications', 'vip-real-time-collaboration' ) }
 					</Heading>
 
-					<Heading level={ 2 } style={ { marginBottom: '2px' } }>
```
Why was this removed?

chriszarate replied:
I thought it was a mistake. It's a level 2 heading following a level 3 and looked anomalous. It's also the only sub-section so provided no distinction.

ingeniumed replied:
It was done on purpose, to allow for the notifications to be categorized based on their purpose. I think the User sub-section was removed? 

There were supposed to be 2 of em:

- Post: This would be for save/publish and any other post specific action
- Collaborator: This would be for when a collaborator enters/exits a post

I think because the second one was removed it looks confusing. Lemme dig that up, and re-add it because the toggles don't make a lot of sense right now. For example, "enters" doesn't quite convey what that means.

ingeniumed replied:
It was removed in https://github.com/Automattic/vip-real-time-collaboration/commit/7eca8cda759b7cf6d8008bf67e4687215ea60475.

I'll add it back in.

shekharnwagh replied:
> It was removed in https://github.com/Automattic/vip-real-time-collaboration/commit/7eca8cda759b7cf6d8008bf67e4687215ea60475.

My bad, I confused it with the new collaborators section which we added in that PR.

chriszarate replied:
Can you fix the heading levels?

ingeniumed replied:
Noting that this has been fixed in #120
